examples/tutorials/pipeline_end_to_end_tutorial.ipynb

```diff
@@ -0,0 +1,1472 @@
+   "id": "8708bddb",
+   "metadata": {},
+   "source": [
+    "#### Module File"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a4301398",
+   "metadata": {},
+   "source": [
+    "Specifically, in the current case, we have `iris_pipeline_module.py` looking as follows:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 19,
+   "id": "3f4e72b1",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Uncomment and run next line to view full file content\n",
+    "# %cat ./pipeline_example/iris_pipeline_module.py"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "acf92dfc",
+   "metadata": {},
+   "source": [
+    "```python\n",
+    "# ./pipeline_example/iris_pipeline_module.py\n",
+    "\n",
+    "[...]\n",
+    "\n",
+    "def get_iris_preprocessed(url):\n",
+    "    df = pd.read_csv(url)\n",
+    "    color_map = {\"Setosa\": \"green\", \"Versicolor\": \"blue\", \"Virginica\": \"red\"}\n",
+    "    df[\"variety_color\"] = df[\"variety\"].map(color_map)\n",
+    "    num_map = {\"Setosa\": 0, \"Versicolor\": 1, \"Virginica\": 2}\n",
+    "    df[\"variety_num\"] = df[\"variety\"].map(num_map)\n",
+    "    return df\n",
+    "\n",
+    "\n",
+    "def get_split_samples_for_artifact_iris_model_and_downstream(\n",
+    "    df, random_state, test_size\n",
+    "):\n",
+    "    X = df[[\"petal.length\", \"petal.width\"]]\n",
+    "    y = df[\"variety_num\"]\n",
+    "    split_samples = train_test_split(\n",
+    "        X, y, test_size=test_size, random_state=random_state\n",
+    "    )\n",
+    "    return split_samples\n",
+    "\n",
+    "\n",
+    "def get_iris_model(split_samples):\n",
+    "    X_train = split_samples[0]\n",
+    "    y_train = split_samples[2]\n",
+    "    mod = LogisticRegression(multi_class=\"multinomial\")\n",
+    "    mod.fit(X_train, y_train)\n",
+    "    return mod\n",
+    "\n",
+    "\n",
+    "def get_iris_model_evaluation(mod, split_samples):\n",
+    "    X_test = split_samples[1]\n",
+    "    y_test = split_samples[3]\n",
+    "    y_test_pred = mod.predict(X_test)\n",
+    "    mod_eval_test = classification_report(y_test, y_test_pred, digits=3)\n",
+    "    return mod_eval_test\n",
+    "\n",
+    "[...]\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2674cf96",
+   "metadata": {},
+   "source": [
+    "We can see that LineaPy used artifacts to automatically 1) clean up their code to retain only essential operations and 2) package the cleaned-up code into importable functions. For instance, we see that `get_iris_preprocessed()` can re-calculate the `iris_preprocessed` artifact, but in a more succinct way (e.g., irrelevant plotting code is now gone).\n",
+    "\n",
+    "Note that pipeline input parameters specified earlier (e.g., `url`) do appear in relevant function signatures (e.g., `def get_iris_preprocessed(url)`), which allows us to pass in different values to run these functions (hence the pipeline).\n",
```

Comment:
"do appear in relevant" is because its parametrized. Let's structure this sentence to use parametrization somehow to describe what is going on here.

Reply:
Good point.